### DISCUSSION TITLE
How to refresh findBy page selectors?

### DISCUSSION THREAD

**Discussion by aybartsch** [serenity-bdd/serenity-core#3479] [Q&A] [answered]
Hi there. I'm mainly using findBy annotation to define web elements on a page like below:
```
    @FindBy(css = "input[name='firstName']")
    protected List<WebElementFacade> inputsFirstName;

    @FindBy(css = "input[name='lastName']")
    protected List<WebElementFacade> inputsLastName;
```

When the page is reloaded in background these references become stale. Is there a way to refresh the references defined above?

I'm aware of dynamic referencing using the $-key but I would prefer to use findBy annotations instead.

Thanks and best regards, André.

**Comment by wakaleo** [comment, marked as answer]
This is a known issue with the PageFactory approach - my solution is never to use @FindBy annotations.

**Comment by aybartsch** [comment]
Thanks, John! Unfortunately, I've used this approach in all my projects because it looked very smart to me in the beginning. What are you suggesting instead? (I'm not using screenplay pattern.)

**Comment by aybartsch** [comment]
@wakaleo Dear John. I'd like to replace the findBy annotation by dynamic selectors but predefining is causing a compiler exception:
```
    protected ListOfWebElementFacades inputsLastName = $$("input[name='lastName']");
    protected ListOfWebElementFacades inputsFirstName = $$("input[name='firstName']");

-->

net.thucydides.core.pages.WrongPageError: The page object class ...RegistrationSearchPage could not be instantiated:
Failed to instantiate page (java.lang.NullPointerException)

	at net.thucydides.core.pages.PageFactory.thisPageObjectLooksDodgy(PageFactory.java:103)
	at net.thucydides.core.pages.PageFactory.createPageOfType(PageFactory.java:50)
```
Could you please advice how to makte it correctly? Thanks and best regards, André

**Comment by aybartsch** [comment]
FYI: I decided to use By instead of FindBy annotation:
```
protected static final By INPUTS_LAST_NAME = By.cssSelector("input[name='lastName']");
protected static final By INPUTS_FIRST_NAME = By.cssSelector("input[name='firstName']");
...
waitAndType($$(INPUTS_LAST_NAME).get(0), decease.getLastName());
waitAndType($$(INPUTS_FIRST_NAME).get(0), decease.getFirstName());
```
